Account Closure: List Sites to be Deleted

Aurorum opened the pull request:
#### Changes proposed in this Pull Request

This lists all sites that get deleted when an account is closed (I believe it's all sites except Jetpack ones? Please correct me if I'm wrong!). There's also a dropdown to hide them, just in case someone has a lot of sites.

#### Testing instructions

Visit `/me/account/close` on an account that can automatically be closed, and verify that the sites display and the toggle works.

**Where no sites will be deleted:**

<img width="890" alt="Screenshot 2019-11-15 at 22 22 01" src="https://user-images.githubusercontent.com/43215253/68981259-c1a11800-07fa-11ea-8e4b-dd2dd93ea026.png">

**Where sites will be deleted:**

<img width="833" alt="Screenshot 2019-11-16 at 09 52 46" src="https://user-images.githubusercontent.com/43215253/68991441-db704880-0856-11ea-9d03-a07bd93a439d.png">


cc @bluefuton 

Fixes #24939

Aurorum commented:
Actually, two cases just occurred to me that need to be tidied up:

- Sites where the user isn't an admin on
- Handling cases where no sites are deleted


I'll fix that up in a bit

Aurorum commented:
Thanks @bluefuton!

I think this is ready for a review now, I've ensured that sites which cannot be deleted don't appear in the list, and if no sites are present, the message says `You have no sites that will be deleted`. :)

bluefuton commented on client/me/account-close/main.jsx:
```diff
@@ -108,7 +127,16 @@ class AccountSettingsClose extends Component {
+									{ this.state.showSiteDropdown && (
+										<SiteSelector
+											filter={ this.siteListFilter }
+											noResultsMessage={ translate( 'You have no sites that will be deleted' ) }
```
If we know we won't be deleting any sites, I wonder if we should just omit the 'Sites' item from the list altogether (along with related ones like posts, pages and media).

Aurorum replied:
I tried to get this working by somehow checking how many sites are filtered, or by trying to check if the `site-selector__no-results` class was on the page, but I couldn't get that working. Any ideas? :)

bluefuton commented:
Thanks for taking this on @Aurorum! :)

Visually, I think we could drop the site avatar to leave more room for the title and URL:

<img width="210" alt="Screen Shot 2019-11-14 at 16 09 44" src="https://user-images.githubusercontent.com/17325/68823657-47d53580-06f9-11ea-8cba-56fbe78b9543.png">

We could probably lose the 'private' badge too, because that's not especially relevant to the user if they're giving up their whole account.

bluefuton commented on client/me/account-close/main.jsx:
```diff
@@ -108,7 +127,16 @@ class AccountSettingsClose extends Component {
+										<Gridicon size={ 18 } onClick={ this.handleSiteDropdown } icon="chevron-down" />
+									</ActionPanelFigureListItem>
+									{ this.state.showSiteDropdown && (
+										<SiteSelector
```
I think this should be inside the `ActionPanelFigureListItem` for the markup to validate:

<img width="584" alt="Screen Shot 2019-11-14 at 16 12 54" src="https://user-images.githubusercontent.com/17325/68823775-a6021880-06f9-11ea-80e6-d3a86c658f18.png">

Aurorum replied:
Will fix when I add a commit to remove the icon! (Otherwise the icon looks a bit odd)

<img width="216" alt="Screenshot 2019-11-14 at 20 55 12" src="https://user-images.githubusercontent.com/43215253/68895400-18d2ba00-0721-11ea-8d8c-bb91884747e3.png">

Aurorum replied:
Done! :)

Aurorum commented:
Thanks for the review! I'll remove the icon/badge soon, but I was wondering if it was better to just use `display: none` or to make a change to the component. I'm not exactly certain since I'm guessing this is the only case where that scenario would ever be necessary, so I'm thinking it's better to just use CSS, is there a strong preference in Calypso?

bluefuton commented:
No worries! This is the last issue in our Account Closure milestone and it's great to see it being worked on.

So - I think the cleanest approach might be to skip the `SiteSelector` component entirely and grab the sites ourselves using the getSites selector (`state/selectors/get-sites`) and associated query component (`<QuerySites allSites />`).

Aurorum commented:
Thanks for setting me on the right direction! I've come up with something which I think is a better approach, and now actually hides things where possible.

**Where no sites will be deleted:**

<img width="890" alt="Screenshot 2019-11-15 at 22 22 01" src="https://user-images.githubusercontent.com/43215253/68981259-c1a11800-07fa-11ea-8e4b-dd2dd93ea026.png">

**Where sites will be deleted:**

<img width="1087" alt="Screenshot 2019-11-15 at 22 41 55" src="https://user-images.githubusercontent.com/43215253/68981270-c960bc80-07fa-11ea-8556-f365591ff6cc.png">

However, I still need to make some changes to how the filter works because of #14024, which I'll probably do at some point tomorrow. I just wanted to commit the changes I've made so far just to verify if the approach & design is okay. :)

Aurorum commented:
Okay, I think this is now ready for review; it appears that the lint error that was appearing previously seems to suggest it's better to create a new selector. I've done so under `getAccountClosureSites` and added some tests, since I think that's the best approach (let me know if not!). 

I've also changed the site domain under the site title to the site slug (`http://cat.com` --> `cat.com`) since it's more consistent with what's usually done in Calypso, and fixed this console error.

<img width="1152" alt="Screenshot 2019-11-16 at 08 30 40" src="https://user-images.githubusercontent.com/43215253/68993912-e2a64f00-0874-11ea-8d83-1ab1943def0a.png">

The only thing I've not been able to fix is the Lint error on line 105, nothing seems to be working, so I've just disabled the Lint rule on that line for now.

Aurorum commented:
Sorry @bluefuton, just weary of the label and wanted to double check if I still needed to change something regarding this PR. :)

bluefuton commented:
Not at all @Aurorum - thanks for the changes! I was on support rotation last week and haven't had a chance to take a look yet. I'll update the label now :)

bluefuton commented on client/me/account-close/main.jsx:
```diff
@@ -81,12 +93,22 @@ class AccountSettingsClose extends Component {
+			'is-hiding-other-sites': this.state.showSiteDropdown,
 		} );
 
+		/* eslint-disable-next-line no-shadow */
```
Can we use a different variable name to avoid skipping the lint check here?

bluefuton commented on client/state/selectors/get-account-closure-sites.js:
```diff
@@ -0,0 +1,19 @@
+/**
```
New selector is a good idea 👍

bluefuton commented on client/state/selectors/test/get-account-closure-sites.js:
```diff
@@ -0,0 +1,96 @@
+/**
+ * External dependencies
+ */
+import { expect } from 'chai';
```
We've moved away from Chai in favour of Jest in newer tests. If you remove the Chai import, you can change your assertions from `to.eql` to `toEqual` and it should work :) You don't need to add any new imports for Jest.

Here's an example test suite:

https://github.com/Automattic/wp-calypso/blob/f2c9d99b07e14ea938829e9b2e5b31b6fb388a90/client/lib/format-number-compact/test/index.js

Aurorum replied:
Thanks, done! :)

bluefuton commented on client/me/account-close/main.jsx:
```diff
@@ -81,12 +93,22 @@ class AccountSettingsClose extends Component {
+
 		return (
 			<div className={ containerClasses } role="main">
 				{ currentUserId && <QueryUserPurchases userId={ currentUserId } /> }
```
I think we should possibly add a `<QuerySites allSites />` just in case we haven't loaded the user's sites when they get to this page (followed a direct link to /me/account/close perhaps)?

jsnajdr replied:
`<QuerySites allSites />` is already done in [client/layout](https://github.com/Automattic/wp-calypso/blob/master/client/layout/index.jsx#L136) for all routes that don't explicitly opt out. It doesn't need to be added here.

Aurorum replied:
For what it's worth, it's also already done in this file.

https://github.com/Automattic/wp-calypso/blob/5670da208eb164bd448d42174c5967a05a25f339/client/me/account-close/main.jsx#L91

bluefuton replied:
Thanks for the information @jsnajdr! In which case, sounds like we can take it out of `me/account-close/index.jsx`.

Aurorum replied:
Done :)

jsnajdr commented:
> I'll remove the icon/badge soon, but I was wondering if it was better to just use display: none or to make a change to the component.

I'm not sure if this question is still relevant ... but it would be better to add a new prop (`showBadge`) to the `<Site>` component. There are already `indicator` and `showHomeLink` props that turn display of other optional info on and off.

jsnajdr commented on client/me/account-close/main.jsx:
```diff
@@ -73,6 +72,18 @@ class AccountSettingsClose extends Component {
+				showSiteDropdown: false,
+			} );
+		}
+	};
```
This code looks weird. If you want to toggle the state flag, use `this.setState( state => ( { showSiteDropdown: ! state.showSiteDropdown } ) )`

Aurorum replied:
Much more elegant! Thanks, done. :)

Aurorum commented:
> I'm not sure if this question is still relevant ... but it would be better to add a new prop (showBadge) to the <Site> component. There are already indicator and showHomeLink props that turn display of other optional info on and off.

Yep, no longer relevant. Initially, I was trying to use a component to display the sites, now it's just a simple list.

bluefuton commented on client/me/account-close/main.jsx:
```diff
@@ -108,10 +113,37 @@ class AccountSettingsClose extends Component {
-									<ActionPanelFigureListItem>{ translate( 'Posts' ) }</ActionPanelFigureListItem>
-									<ActionPanelFigureListItem>{ translate( 'Pages' ) }</ActionPanelFigureListItem>
-									<ActionPanelFigureListItem>{ translate( 'Media' ) }</ActionPanelFigureListItem>
+									{ this.props.siteToBeDeleted.length > 0 && (
```
Can we change this to the plural `sitesToBeDeleted`?

bluefuton commented on client/me/account-close/main.jsx:
```diff
@@ -256,5 +288,6 @@ export default connect( state => {
 		purchasedPremiumThemes,
 		hasAtomicSites: userHasAnyAtomicSites( state ),
 		isAccountClosed: isAccountClosed( state ),
+		siteToBeDeleted: getAccountClosureSites( state ),
```
Can we change this to the plural `sitesToBeDeleted`?

bluefuton commented on client/me/account-close/style.scss:
```diff
@@ -11,6 +11,39 @@
 }
 
+.account-close {
+	.account-close__sites-item {
```
As they're "namespaced" already with `account-close__`, I think we can move these style definitions outside of the `account-close` block (i.e. move `account-close__sites-item` and `account-close__sites-list` to the top-level).

bluefuton commented:
Thank you for all your work on this @Aurorum!

I left a few tiny nits, but it tests out well. If we get those small things fixed we can ship it 🚢

Aurorum commented:
Thanks @bluefuton, I've updated the PR to address those issues! :)
